lib/axon/losses.ex

```diff
@@ -1024,4 +1024,170 @@ defmodule Axon.Losses do
   end
+
+  @doc """
+  WIP
+  y_pred should be log_softmax output of shape (B, T, D),
+  y_true should be of shape (B, S),
+  where
+  B - batch size,
+  T - sequence length,
+  D - dictionary size, including blank
+  S - maximum target length
+  y_true nas to be preencoded, its first and last characters should be
+  blank, each initial character should be separated with blanks,
+  double blanks mark the end of target sequence.
+  """
+  defn ctcloss(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, blank: 0, reduction: :none)
+    # eps = Nx.tensor([1.0e-7])
+    b_size = elem(Nx.shape(y_true), 0)
+    t_max = elem(Nx.shape(y_pred), 1) - 1
+    loss = Nx.broadcast(0.0, {b_size})
+
+    {loss, _, _, _} =
+      while {loss, b = 0, y_true, y_pred}, b < b_size do
+        # Get boundaries for available node paths.
+        st_lims = get_limits(y_true[b], t_max, opts[:blank])
+        # Iterate node tree backwards.
+        s_pred0 = iterate_tree(y_true[b], y_pred[b], st_lims, t_max)
+
+        {loss_b, _, _} =
+          while {loss_b = 0.0, s = st_lims[0][0], s_pred0}, s <= st_lims[0][1] do
+            {Nx.add(loss_b, Nx.exp(s_pred0[s])), s + 1, s_pred0}
+          end
+
+        loss_b = -Nx.log(loss_b)
+        {Nx.put_slice(loss, [b], Nx.reshape(loss_b, {1})), b + 1, y_true, y_pred}
+      end
+
+    transform(
+      {opts[:reduction], loss},
+      fn
+        # {:mean, loss} -> Nx.mean(loss)
+        {:sum, loss} -> Nx.sum(loss)
+        {:none, loss} -> loss
+      end
+    )
+  end
+
+  defn get_limits(y_true, t_max, blank) do
+    y_true_bb = Nx.concatenate([y_true, Nx.reshape(blank, {1})])
+    # Get `trimmed` tagret length.
+    {s_max, _} =
+      while {s = 0, y_true_bb}, y_true_bb[s] != y_true_bb[s + 1] do
+        {s + 1, y_true_bb}
+      end
+
+    st_max = Nx.concatenate([Nx.tensor([1]), Nx.broadcast(s_max, {t_max})])
+    # Iterate target to get upper boundary values for each sequence step.
+    {st_max, _, t_fin, _} =
+      while {st_max, s = 1, t = 1, y_true}, t <= t_max and s <= s_max - 2 do
+        s =
+          cond do
+            y_true[s] != y_true[s + 2] -> s + 2
+            true -> s + 1
+          end
+
+        {Nx.put_slice(st_max, [t], Nx.reshape(s, {1})), s, t + 1, y_true}
+      end
+
+    # transform({s_fin, s_max}, fn {s_fin, s_max} ->
+    # unless Elixir.Kernel.==(s_fin, s_max - 1) do
+    # raise ArgumentError, 
+    # "either prediction sequence is too short for label sequence" <>
+    # " or label sequence has double blank characters"
+    # end
+    # end)
+    # Get lower boundary values based on upper boundary values.
+    st_min =
+      cond do
+        t_fin == t_max + 1 ->
+          st_max
+
+        true ->
+          st_min = Nx.broadcast(0, {t_max + 1})
+
+          {st_min, _, _} =
+            while {st_min, dt = 1, st_max}, dt <= t_fin do
+              {Nx.put_slice(st_min, [t_max - dt + 1], Nx.reshape(st_max[t_fin - dt], {1})),
+               dt + 1, st_max}
+            end
```

Comment:
Does this need to be sequential?

Reply:
I think it does. It's recursive by nature.

Reply:
I can't use `Nx.put_slice` in one step, because I can't use `t_fin` in slice length, as long as it's scalar type, I think.

Reply:
I think I know how to get rid of it. Will try.

Reply:
Awesome! The more while loops we can get rid of the more performant this will be.

Reply:
Didn't work out. Sorry.